File: py/vcf/predictor.py
from abc import ABCMeta, abstractmethod
from collections import defaultdict, Counter
from typing import Dict, List, TypeVar, Tuple, cast
from functools import lru_cache
from itertools import combinations

T = TypeVar('T')
V = TypeVar('V')

    
class Predictor(metaclass=ABCMeta):
    @abstractmethod
    def predict(self, genotype: float) -> Dict[str, float]:
        pass


class FrequencyPredictor(Predictor, metaclass=ABCMeta):
    
    def __init__(self, genotypes: List[float], labels: List[str]) -> None:
        counts: Dict[str, float] = defaultdict(float)
        totals: Dict[str, int] = defaultdict(int)
        
        for label, frequency in zip(labels, genotypes):
            counts[label] += frequency
            totals[label] += 1
        
        self.frequencies = {
            label: count / totals[label]
            for label, count in counts.items()
        }


class PredictorJoao(FrequencyPredictor):
    
    def __init__(self, genotypes: List[float], labels: List[str], threshold: float) -> None:
        super().__init__(genotypes, labels)
        self.threshold = threshold
    
    @lru_cache(maxsize=16)
    def predict(self, genotype: float) -> Dict[str, float]:
        distances = {
            label: abs(genotype - freq)
            for label, freq in self.frequencies.items()
        }
        ordered = sorted(distances, key=lambda x: distances[x])
        if distances[ordered[1]] - distances[ordered[0]] >= self.threshold:
            return {ordered[0]: 1.0}
        else:
            return {}


class PredictorMariana(FrequencyPredictor):
    
    def __init__(self, genotypes: List[float], labels: List[str], threshold: float) -> None:
        super().__init__(genotypes, labels)
        self.threshold = threshold
    
    @lru_cache(maxsize=16)
    def predict(self, genotype: float) -> Dict[str, float]:
        predictions = []
        pairs = combinations(self.frequencies, 2)
        for p1, p2 in pairs:
            d = abs(genotype - self.frequencies[p1]) - abs(genotype - self.frequencies[p2])
            if abs(d) >= self.threshold:
                if d < 0:
                    predictions.append(p1)
                else:
                    predictions.append(p2)
        
        counts = Counter(predictions)
        size = len(predictions)
        return {key: count / size for key, count in counts.items()}


class PredictorFactory(metaclass=ABCMeta):
    
    @abstractmethod
    def build(self, genotypes: List[float], labels: List[str]) -> Predictor:
        pass


class PredictorJoaoFactory(PredictorFactory):
    
    def __init__(self, threshold: float) -> None:
        super().__init__()
        self.threshold = threshold
    
    def build(self, genotypes: List[float], labels: List[str]) -> Predictor:
        return PredictorJoao(genotypes, labels, self.threshold)


class PredictorMarianaFactory(PredictorFactory):
    
    def __init__(self, threshold: float) -> None:
        super().__init__()
        self.threshold = threshold
    
    def build(self, genotypes: List[float], labels: List[str]) -> Predictor:
        return PredictorMariana(genotypes, labels, self.threshold)

File: rs/src/vcf/predictor.rs
use std::collections::HashMap;

use itertools::Itertools;

use utils::compute_frequencies;

pub trait PredictorFactory
{
    type Predictor: Predictor;
    
    fn build<I>(&self, genotypes_and_labels: I) -> Self::Predictor
    where
        I: Iterator<Item = (<Self::Predictor as Predictor>::Prediction, String)>;
}

pub trait Predictor {
    type Prediction;
    
    fn predict(&self, genotype: Self::Prediction) -> HashMap<&str, f32>;
}

pub struct PredictorJoao {
    threshold: f32,
    frequencies: HashMap<String, f32>,
}

impl Predictor for PredictorJoao {
    type Prediction = f32;
    
    fn predict(&self, genotype: f32) -> HashMap<&str, f32> {
        let distances = self
            .frequencies
            .iter()
            .map(|(label, freq)| (&**label, f32::abs(genotype - freq)))
            .sorted_by(|(_, f1), (_, f2)| f1.partial_cmp(f2).unwrap());

        if distances[1].1 - distances[0].1 >= self.threshold {
            vec![(distances[0].0, 1.0)].iter().cloned().collect()
        }
        else {
            vec![].iter().cloned().collect()
        }
    }
}

pub struct PredictorJoaoFactory {
    pub threshold: f32,
}

impl PredictorFactory for PredictorJoaoFactory {
    type Predictor = PredictorJoao;
    
    fn build<I>(&self, genotypes_and_labels: I) -> PredictorJoao
    where
        I: Iterator<Item = (f32, String)>,
    {
        PredictorJoao {
            threshold: self.threshold,
            frequencies: compute_frequencies(genotypes_and_labels),
        }
    }
}
